### PR TITLE
Reduce requirements for parameter in catalog product type factory

### PR DESCRIPTION
**Pull request by hws47a**
### Description (*)
In the factory method, only API Interface method is used, but PHPDoc asks for the Product Model instead. Fix for this.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Review by novikor — CHANGES_REQUESTED** [review]
Hi, @hws47a .
Thank you for your contribution.
Except the changes provided, I think it would be nice to execute some code cleanup - at least replace fully qualified class names with import, and probably add missing `@throws` tags.
image

**Comment by hws47a** [comment]
Hi @novikor, thanks for your review. 

My idea is opposite, I do small  changes but which bother me. 
For example in this case I won't need to a check in a level above for   
`$product instance of \Magento\Catalog\Model\Product`   
after loading product from repository, which will simplify the code for me.

I believe in every single core class there are plenty of code which could be cleaned up, but it'll take ages if I'll start.

So please accept the PR as it is if it passes checks.

**Review by novikor — APPROVED** [review]
@hws47a, As you wish

**Comment by magento-engcom-team** [comment]
Hi @novikor, thank you for the review.
[ENGCOM-6883](https://jira.corp.magento.com/browse/ENGCOM-6883) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Comment by engcom-Golf** [comment]
Failed functional tests not related to the changes in this PR.
SVC fail is not valid as only one type was changed, need to approve SVC failure.

**Comment by novikor** [comment]
Hmm. There were no SVC failures at the moment of approval.
2 new commits created after caused that.

**Comment by novikor** [comment]
@Nazar65, please do not add new commits to PRs which are already approved next time :)

**Comment by novikor** [comment]
> Failed functional tests not related to the changes in this PR.
> SVC fail is not valid as only one type was changed, need to approve SVC failure.

It seems that SVC false alarm happened. There are no methods return types changes by the fact.

**Comment by hws47a** [comment]
Thanks @novikor

**Comment by slavvka** [comment]
The ticket to SVC approval https://jira.corp.magento.com/browse/MC-31723

File: app/code/Magento/Catalog/Model/Product/Type.php [hunk 1]
```diff
@@ -164,7 +161,7 @@ public function priceFactory($productType)
      * Get Product Price Info object
      *
      * @param Product $saleableItem
-     * @return \Magento\Framework\Pricing\PriceInfoInterface
```
Comment (naydav) [inline app/code/Magento/Catalog/Model/Product/Type.php]: Changing of input/return annotations in docblocks should be reverted (for **all methods** except construct).
Theoretically, this class can be used in WebAPI requests.

The current implementation of WebAPI "request/response" parser has a limitation:

> Valid object types include a fully qualified class name or a fully qualified interface name.
https://devdocs.magento.com/guides/v2.3/extension-dev-guide/service-contracts/service-to-web-service.html

Comment (slavvka) [reply app/code/Magento/Catalog/Model/Product/Type.php]: @hws47a Please make changes according to @naydav comment.

**Review by slavvka — DISMISSED** [review]
Please fix the PR according to @naydav comment

File: app/code/Magento/Catalog/Model/Product/Type.php [hunk 2]
```diff
@@ -116,8 +113,8 @@ public function __construct(
     /**
      * Factory to product singleton product type instances
      *
-     * @param   \Magento\Catalog\Model\Product $product
-     * @return  \Magento\Catalog\Model\Product\Type\AbstractType
+     * @param ProductInterface $product
```
Comment (engcom-Golf) [inline app/code/Magento/Catalog/Model/Product/Type.php]: @slavvka is this type also might be reverted ? or we can leave `Magento\Catalog\Api\Data\ProductInterface`?

Comment (slavvka) [reply app/code/Magento/Catalog/Model/Product/Type.php]: Use full qualified type for ProductInterface. Don't import it.

Comment (engcom-Golf) [reply app/code/Magento/Catalog/Model/Product/Type.php]: ok, i mean that in this case return type was changed from  `\Magento\Catalog\Model\Product `to `Magento\Catalog\Api\Data\ProductInterface`

Comment (slavvka) [reply app/code/Magento/Catalog/Model/Product/Type.php]: Exactly, without it there's no reason for making this PR

Comment (engcom-Golf) [reply app/code/Magento/Catalog/Model/Product/Type.php]: done :heavy_check_mark:

**Comment by engcom-Golf** [comment]
Hi @novikor could you please review this one again ?

**Review by novikor — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @novikor, thank you for the review.
[ENGCOM-6883](https://jira.corp.magento.com/browse/ENGCOM-6883) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Pull request merged**